src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+/* @flow */
+import { createWriteStream, promises as fsPromises } from 'fs';
+import { pipeline } from 'stream';
+import { promisify } from 'util';
+
+// eslint-disable-next-line no-shadow
+import fetch, { FormData, fileFromSync, Response } from 'node-fetch';
+import defaultJwt from 'jsonwebtoken';
+
+export type SignResult = {|
+  success: boolean,
+  id: string | null,
+  downloadedFiles: Array<string>,
+|};
+
+type ClientConstructorParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiUrlPrefix: string,
+  apiJwtExpiresIn?: number,
+  debugLogging?: boolean,
+  validationCheckInterval?: number,
+  validationCheckTimeout?: number,
+  approvalCheckInterval?: number,
+  approvalCheckTimeout?: number,
+  logger?: any,
+  downloadDir?: string,
+|};
+
+export default class Client {
+  apiKey: string;
+  apiSecret: string;
+  apiUrlPrefix: string;
+  apiJwtExpiresIn: number;
+  debugLogging: boolean;
+  validationCheckInterval: number;
+  validationCheckTimeout: number;
+  approvalCheckInterval: number;
+  approvalCheckTimeout: number;
+  logger: any;
+  downloadDir: string;
+
+  constructor({
+    apiKey,
+    apiSecret,
+    apiUrlPrefix,
+    apiJwtExpiresIn = 60 * 5, // 5 minutes
+    debugLogging = false,
+    validationCheckInterval = 1000,
+    validationCheckTimeout = 300000, // 5 minutes.
+    approvalCheckInterval = 1000,
+    approvalCheckTimeout = 900000, // 15 minutes.
+    logger = console,
+    downloadDir = process.cwd(),
+  }: ClientConstructorParams) {
+    this.apiKey = apiKey;
+    this.apiSecret = apiSecret;
+    this.apiUrlPrefix = `${apiUrlPrefix}addons/`;
+    this.apiJwtExpiresIn = apiJwtExpiresIn;
+    this.validationCheckInterval = validationCheckInterval;
+    this.validationCheckTimeout = validationCheckTimeout;
+    this.approvalCheckInterval = approvalCheckInterval;
+    this.approvalCheckTimeout = approvalCheckTimeout;
+    this.debugLogging = debugLogging;
+    this.logger = logger;
+    this.downloadDir = downloadDir;
+  }
+
+  fileFromSync(path: string): File {
+    return fileFromSync(path);
+  }
+
+  async doUploadSubmit(xpiPath: string, channel: string): Promise<string> {
+    const url = `${this.apiUrlPrefix}upload/`;
+    const formData = new FormData();
+    formData.set('channel', channel);
+    formData.set('upload', this.fileFromSync(xpiPath));
+    const { uuid } = await this.getJson(
+      await this.fetch(url, 'POST', formData)
+    );
+    return this.waitForValidation(uuid);
+  }
+
+  waitForValidation(uuid: string): Promise<string> {
+    let validationCheckTimeout;
+
+    return new Promise((resolve, reject) => {
+      const uploadDetailUrl = `${this.apiUrlPrefix}upload/${uuid}/`;
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(validationCheckTimeout);
+
+        reject(new Error('Validation Timeout.'));
+      }, this.validationCheckTimeout);
```

Comment:
> Our CI job runs on nodejs 14.19 and so it would also work without the polyfill, but the web-ext package is declared as compatible with nodejs >= 14 and so we would need the polyfill dependency to keep all nodejs 14 version as supported if we plan to use the AbortController.

I don't think we need a polyfill. 14.x is a maintenance LTS and it got security issues not too long ago: https://nodejs.org/en/blog/vulnerability/july-2022-security-releases/, so it is important to not rely on older 14.x versions.

I think we should update the version `package.json` instead.

As for `AbortController`, yeah that looks good.

Reply:
I can add this `AbortController`, but I don't see how it turns the function into something more visually linear.  From the link it appears to be _extra_ lines of code that abort the fetch if the abort timeout happens mid-fetch.

(And because the fetch call is 2 functions down, via `this.fetchJson` and `this.fetch`, passing down an optional `AbortController` adds additional complexity to those functions too)

Reply:
yeah, after starting adding `AbortController`, if the aim is more readable code, then it doesn't achieve the aim, imo.

Reply:
https://github.com/mozilla/web-ext/issues/2476